lib/url.js

```diff
@@ -75,355 +88,192 @@ const slashedProtocol = {
+  '#',
+  '?',
+  '/',
+  '\\'
+].map(function(v) {
+  return v.charCodeAt(0);
+}));
+// If these characters end a host name, the path will not be prepended a /.
+const _hostEndingCharactersNoPrependSlash = makeAsciiTable([
+  '<',
+  '>',
+  '"',
+  '`',
+  ' ',
+  '\r',
+  '\n',
+  '\t',
+  '{',
+  '}',
+  '|',
+  '^',
+  '`',
+  '\'',
+  '%',
+  ';'
+].map(function(v) {
+  return v.charCodeAt(0);
+}));
+
 const querystring = require('querystring');
 
+const noSerializePattern = /^(?:_protocol|_port|_href|_query|_prependSlash)$/;
+
 function urlParse(url, parseQueryString, slashesDenoteHost) {
   if (url instanceof Url) return url;
 
-  var u = new Url;
+  var u = new Url();
   u.parse(url, parseQueryString, slashesDenoteHost);
   return u;
 }
 
-Url.prototype.parse = function(url, parseQueryString, slashesDenoteHost) {
-  if (typeof url !== 'string') {
-    throw new TypeError("Parameter 'url' must be a string, not " + typeof url);
-  }
-
-  // Copy chrome, IE, opera backslash-handling behavior.
-  // Back slashes before the query string get converted to forward slashes
-  // See: https://code.google.com/p/chromium/issues/detail?id=25916
-  var queryIndex = url.indexOf('?'),
-      splitter =
-          (queryIndex !== -1 && queryIndex < url.indexOf('#')) ? '?' : '#',
-      uSplit = url.split(splitter),
-      slashRegex = /\\/g;
-  uSplit[0] = uSplit[0].replace(slashRegex, '/');
-  url = uSplit.join(splitter);
-
-  var rest = url;
-
-  // trim before proceeding.
-  // This is to support parse stuff like "  http://foo.com  \n"
-  rest = rest.trim();
-
-  if (!slashesDenoteHost && url.split('#').length === 1) {
-    // Try fast path regexp
-    var simplePath = simplePathPattern.exec(rest);
-    if (simplePath) {
-      this.path = rest;
-      this.href = rest;
-      this.pathname = simplePath[1];
-      if (simplePath[2]) {
-        this.search = simplePath[2];
-        if (parseQueryString) {
-          this.query = querystring.parse(this.search.substr(1));
-        } else {
-          this.query = this.search.substr(1);
-        }
-      } else if (parseQueryString) {
-        this.search = '';
-        this.query = {};
-      }
-      return this;
-    }
+Url.prototype.parse = function(str, parseQueryString, hostDenotesSlash) {
+  if (typeof str !== 'string') {
+    throw new TypeError(`Parameter 'url' must be a string, not ` +
+        typeof str);
   }
-
-  var proto = protocolPattern.exec(rest);
-  if (proto) {
-    proto = proto[0];
-    var lowerProto = proto.toLowerCase();
-    this.protocol = lowerProto;
-    rest = rest.substr(proto.length);
+  var start = 0;
+  var end = str.length - 1;
+
+  // Trim leading and trailing ws.
+  while (str.charCodeAt(start) <= 0x20 /*' '*/) start++;
+  var trimmedStart = start;
+  while (str.charCodeAt(end) <= 0x20 /*' '*/) end--;
+
+  start = this._parseProtocol(str, start, end);
+
+  // Javascript doesn't have host.
+  if (this._protocol !== 'javascript') {
```

Comment:
Yes, I was just pointing out that the spec has always been pretty clear
about file:// having a host. The more important reference point, however,
are @rubys test cases showing the differences across various tools.
On Apr 15, 2015 2:15 PM, "Domenic Denicola" <notifications@github.com>
wrote:

> In lib/url.js
> <https://github.com/iojs/io.js/pull/933#discussion_r28463776>:
>
> > -    proto = proto[0];
> > -    var lowerProto = proto.toLowerCase();
> > -    this.protocol = lowerProto;
> > -    rest = rest.substr(proto.length);
> > +  var start = 0;
> > +  var end = str.length - 1;
> > +
> > +  // Trim leading and trailing ws.
> > +  while (str.charCodeAt(start) <= 0x20 /*' '*/) start++;
> > +  var trimmedStart = start;
> > +  while (str.charCodeAt(end) <= 0x20 /*' '*/) end--;
> > +
> > +  start = this._parseProtocol(str, start, end);
> > +
> > +  // Javascript doesn't have host.
> > +  if (this._protocol !== 'javascript') {
>
> io.js aims for compliance to the URL Standard, not the RFCs it obsoletes.
> _____________________________ From:James M Snell Subject:Re: [io.js] url:
> significantly improve the performance of the url module (#933
> <https://github.com/iojs/io.js/pull/933>) To:iojs/io.js Cc:Domenic
> Denicola In lib/url.js<
> https://github.com/iojs/io.js/pull/933#discussion_r28462922>:
>  - proto = proto[0]; - var lowerProto = proto.toLowerCase(); -
> this.protocol = lowerProto; - rest = rest.substr(proto.length); + var start
> = 0; + var end = str.length - 1; + + // Trim leading and trailing ws. +
> while (str.charCodeAt(start) <= 0x20 /*' '*/) start++; + var trimmedStart =
> start; + while (str.charCodeAt(end) <= 0x20 /*' '*/) end--; + + start =
> this._parseProtocol(str, start, end); + + // Javascript doesn't have host.
> + if (this._protocol !== 'javascript') {
> Spec says this: https://tools.ietf.org/html/draft-hoffman-file-uri-03,
> some of this work may come in handy too:
> http://intertwingly.net/blog/2014/10/02/WHATWG-URL-vs-IETF-URI — Reply to
> this email directly or view it on GitHub<
> https://github.com/iojs/io.js/pull/933/files#r28462922>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/iojs/io.js/pull/933/files#r28463776>.
>